File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/booking/RideBookingScreen.kt
package lk.mageride.passenger.booking

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material3.AssistChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.map.MageRideMap
import lk.mageride.passenger.map.MapCamera
import lk.mageride.passenger.map.MapPin
import lk.mageride.passenger.map.VehicleLayers
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ride.RidePaymentMethod
import org.koin.androidx.compose.koinViewModel

/**
 * SCR-PA-009 — the multimodal list, and the primary screen of cluster 3.
 *
 * The wireframe: a map with the route and walk lines over a sheet carrying the pickup/drop summary,
 * the two toggles, the **Public buses · direct routes (GTFS)** section, the **Private (Mode C ·
 * standby)** tiers, and one CTA that says *"Track Route"* or *"Book Now"* depending on which list
 * was chosen from.
 *
 * **Two fences are visible in this file.**
 *
 * - **AL-19.** [TierRow] renders `quote.amountMinor` and there is nothing else on a [TierQuote] to
 *   render — no ETA, no distance. The type is the enforcement; this composable simply cannot show
 *   what it was not given.
 * - **AL-18.** A public row shows route number, description and the Direct/Transit tag, and
 *   selecting one hides the payment chip entirely, because no fare is charged for a bus.
 */
@Composable
@Suppress("LongMethod") // The wireframe's layout tree: map, summary, toggles, two lists, CTA.
internal fun RideBookingScreen(
    onBack: () -> Unit,
    onEditRoute: () -> Unit,
    onProxyDetails: () -> Unit,
    onPackageBooking: () -> Unit,
    onSchedule: () -> Unit,
    onBooked: (String) -> Unit,
    onTrackRoute: () -> Unit,
    model: RideBookingViewModel = koinViewModel(),
) {
    val state by model.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.booked) {
        state.booked?.let {
            onBooked(it)
            model.onBookingConsumed()
        }
    }

    Column(modifier = Modifier.fillMaxSize()) {
        Box(modifier = Modifier.weight(MAP_WEIGHT)) {
            MageRideMap(
                modifier = Modifier.fillMaxSize(),
                routePolyline = state.routePolyline,
                walkPolyline = state.walkPolyline,
                pins = listOfNotNull(
                    state.draft.pickup?.let { MapPin(VehicleLayers.PIN_PICKUP, it.lat, it.lng) },
                    state.draft.dropoff?.let { MapPin(VehicleLayers.PIN_DROPOFF, it.lat, it.lng) },
                ),
                camera = state.draft.pickup?.let { MapCamera(it.lat, it.lng) } ?: MapCamera.Default,
            )
            IconButton(onClick = onBack, modifier = Modifier.align(Alignment.TopStart)) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurface,
                )
            }
        }

        Column(
            modifier = Modifier
                .weight(SHEET_WEIGHT)
                .fillMaxWidth()
                .background(
                    MaterialTheme.colorScheme.background,
                    RoundedCornerShape(
                        topStart = MageRideTheme.radius.card,
                        topEnd = MageRideTheme.radius.card,
                    ),
                )
                .verticalScroll(rememberScrollState())
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            JourneySummary(
                pickup = state.draft.pickup?.address,
                dropoff = state.draft.dropoff?.address,
                onEdit = onEditRoute,
            )

            Toggles(
                state = state,
                onBookingFor = { value ->
                    model.setBookingFor(value)
                    if (value == BookingFor.SOMEONE_ELSE) onProxyDetails()
                },
                onSubject = { value ->
                    model.setSubject(value)
                    if (value == BookingSubject.PACKAGE) onPackageBooking()
                },
            )

            PublicSection(state = state, onSelect = model::selectRoute)

            SectionLabel(text = stringResource(R.string.booking_private_section))
            when {
                state.tiersLoading -> Loading(stringResource(R.string.booking_estimating))

                state.tiers.isEmpty() -> MutedRow(stringResource(R.string.booking_no_tiers))

                else -> state.tiers.forEach { quote ->
                    TierRow(
                        quote = quote,
                        selected = (state.selection as? BookingSelection.Private)?.quote == quote,
                        onSelect = { model.selectTier(quote) },
                    )
                }
            }

            // "Payment chip … (private tiers only)". A bus is not paid for in this app, so the
            // chip is not merely disabled on a public selection — it is absent.
            if (!state.isPublicSelected) {
                PaymentChip(method = state.draft.paymentMethod, onChange = model::setPaymentMethod)
            }

            state.error?.let { InlineError(message = stringResource(it)) }

            BookingCta(
                state = state,
                onBook = model::book,
                onSchedule = onSchedule,
                onTrackRoute = onTrackRoute,
            )
        }
    }
}

/**
 * The wireframe's `● Galle Face ✎ ◆ Nugegoda` card — both ends, each one named.
 *
 * **The two ends are labelled and dotted, and that is a defect fix rather than decoration.** This
 * was two bare lines of text: the pickup above the drop-off, same colour, one type step apart and
 * no marker on either. A pickup with no street address prints *"Current location"* — which
 * `LastKnownFix` is right to do, since the fix is a coordinate nobody has geocoded — and it is
 * ALWAYS what this row shows, because nothing in the app writes an address onto a pickup. So every
 * passenger who chose a destination read a small grey *"Current location"* sitting directly on top
 * of the place they were travelling to, and read it as that place's caption. Reported from a
 * handset. The wireframe's own answer is the coloured `cdot` on each row; the labels are the rest
 * of it.
 *
 * **The ✎ is on the destination row**, because [onEdit] is what sets
 * [CaptureTarget.BOOKING_DROPOFF] — it opens SCR-PA-008 to change where the journey ENDS. It used
 * to sit at the end of a Row whose whole surface was clickable, so a tap on the pickup line opened
 * a picker that then rewrote the drop-off.
 */
@Composable
private fun JourneySummary(pickup: String?, dropoff: String?, onEdit: () -> Unit) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
    ) {
        SectionLabel(text = stringResource(R.string.booking_pickup_label))
        JourneyEnd(
            dot = MageRideTheme.status.success,
            text = pickup ?: stringResource(R.string.search_current_location),
        )
        SectionLabel(
            text = stringResource(R.string.booking_destination_label),
            modifier = Modifier.padding(top = MageRideTheme.spacing.xxs),
        )
        JourneyEnd(
            dot = MaterialTheme.colorScheme.error,
            text = dropoff ?: stringResource(R.string.booking_no_destination),
            onEdit = onEdit,
        )
    }
}

/**
 * One end of the journey: the wireframe's coloured `cdot`, the address, and the ✎ on the end that
 * can be changed.
 *
 * @param onEdit `null` on an end this screen cannot edit, which is what leaves the row without a
 *   click target as well as without the icon — an affordance and its action arrive together.
 */
@Composable
private fun JourneyEnd(dot: Color, text: String, onEdit: (() -> Unit)? = null) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .then(if (onEdit == null) Modifier else Modifier.clickable(onClick = onEdit)),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Box(
            modifier = Modifier
                .size(MageRideTheme.spacing.sm)
                .background(dot, CircleShape),
        )
        Text(
            text = text,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurface,
        )
        if (onEdit != null) {
            Icon(
                imageVector = Icons.Filled.Edit,
                contentDescription = stringResource(R.string.booking_edit_route),
                modifier = Modifier.size(ControlTokens.RowIcon),
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

/** The wireframe's `[For Me | Someone else] [Person | Package]`. */
@Composable
private fun Toggles(state: RideBookingState, onBookingFor: (BookingFor) -> Unit, onSubject: (BookingSubject) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        ToggleChip(
            label = stringResource(R.string.booking_for_me),
            selected = state.draft.bookingFor == BookingFor.ME,
            onClick = { onBookingFor(BookingFor.ME) },
        )
        ToggleChip(
            label = stringResource(R.string.booking_for_someone),
            selected = state.draft.bookingFor == BookingFor.SOMEONE_ELSE,
            onClick = { onBookingFor(BookingFor.SOMEONE_ELSE) },
        )
        ToggleChip(
            label = stringResource(R.string.booking_person),
            selected = state.draft.subject == BookingSubject.PERSON,
            onClick = { onSubject(BookingSubject.PERSON) },
        )
        ToggleChip(
            label = stringResource(R.string.booking_package),
            selected = state.draft.subject == BookingSubject.PACKAGE,
            onClick = { onSubject(BookingSubject.PACKAGE) },
        )
    }
}

@Composable
internal fun ToggleChip(label: String, selected: Boolean, onClick: () -> Unit) {
    AssistChip(
        onClick = onClick,
        label = { Text(label) },
        colors = androidx.compose.material3.AssistChipDefaults.assistChipColors(
            containerColor = if (selected) {
                MaterialTheme.colorScheme.primaryContainer
            } else {
                MaterialTheme.colorScheme.surface
            },
        ),
    )
}

/**
 * The one CTA, in whichever of its two forms applies.
 *
 * *"public → Track Route · private → Book Now / Schedule"* is D2' §SCR-PA-009's own sketch. There is
 * no Schedule under a public route because there is nothing to schedule: a bus runs to its own
 * timetable and this app does not book a seat on it.
 */
@Composable
private fun BookingCta(state: RideBookingState, onBook: () -> Unit, onSchedule: () -> Unit, onTrackRoute: () -> Unit) {
    if (state.isPublicSelected) {
        MageRideCta(label = stringResource(R.string.booking_track_route), onClick = onTrackRoute)
        return
    }

    MageRideCta(
        label = stringResource(R.string.booking_book_now),
        onClick = onBook,
        enabled = state.canBook,
        loading = state.booking,
    )
    OutlinedButton(onClick = onSchedule, modifier = Modifier.fillMaxWidth()) {
        Text(stringResource(R.string.booking_schedule))
    }
}

/** The map is the backdrop and the sheet is the work — the wireframe's own proportions. */
private const val MAP_WEIGHT = 0.4f
private const val SHEET_WEIGHT = 0.6f

File: apps/passenger-ios/PassengerApp/Booking/RideBookingScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-009 — the multimodal list, and the booking.
///
/// The cell: a short map carrying the selected route's polyline, the blue walk-to-halt leg and the
/// two pins, over a scrolling sheet with the journey summary, the two segmented pickers, the GTFS
/// section, the Mode C tiers and one CTA.
///
/// **Two fences are visible in the layout.** A public row shows a route number, a description and a
/// Direct/Transit tag with a `PUBLIC` badge — never a price, because a bus is not booked here; a
/// private tier shows a price and nothing else, because no driver has been matched (AL-19). Choosing
/// a public route also **removes the payment chip entirely** rather than disabling it: there is no
/// fare to choose a rail for.
///
/// **Δ the cell draws no navigation bar**, so this screen hides the system's and puts the back
/// control on the map, which is also what `apps/passenger-android` does. A pushed SwiftUI screen
/// keeps its edge-swipe either way; what the drawn button adds is a target the wireframe's own frame
/// has room for and a `.toolbar` would take 44pt of map to provide.
@MainActor
struct RideBookingScreen: View {

    @StateObject private var model: RideBookingModel

    let onBack: () -> Void
    let onEditRoute: () -> Void
    let onProxyDetails: () -> Void
    let onPackageBooking: () -> Void
    let onSchedule: () -> Void
    let onBooked: (String) -> Void
    let onTrackRoute: () -> Void

    init(
        draft: BookingDraft,
        bookings: BookingRepository,
        keys: IdempotencyKeys,
        onBack: @escaping () -> Void,
        onEditRoute: @escaping () -> Void,
        onProxyDetails: @escaping () -> Void,
        onPackageBooking: @escaping () -> Void,
        onSchedule: @escaping () -> Void,
        onBooked: @escaping (String) -> Void,
        onTrackRoute: @escaping () -> Void
    ) {
        _model = StateObject(wrappedValue: RideBookingModel(draft: draft, bookings: bookings, keys: keys))
        self.onBack = onBack
        self.onEditRoute = onEditRoute
        self.onProxyDetails = onProxyDetails
        self.onPackageBooking = onPackageBooking
        self.onSchedule = onSchedule
        self.onBooked = onBooked
        self.onTrackRoute = onTrackRoute
    }

    var body: some View {
        VStack(spacing: 0) {
            map
            sheet
        }
        .background(MageRideColor.surface)
        .toolbar(.hidden, for: .navigationBar)
        .task { model.start() }
        .onChange(of: model.state.booked) { rideId in
            guard let rideId else { return }
            onBooked(rideId)
            model.onBookingConsumed()
        }
    }

    private var map: some View {
        MageRideMap(
            routePolyline: model.state.routePolyline,
            walkPolyline: model.state.walkPolyline,
            pins: model.state.pins,
            camera: model.state.draft.pickup.map { MapCamera(lat: $0.lat, lng: $0.lng) } ?? .colombo
        )
        .frame(height: MageRideControl.bookingMapHeight)
        .overlay(alignment: .topLeading) {
            MapOverlayButton(symbolName: "chevron.left", accessibilityKey: "action_back", action: onBack)
                .padding(MageRideSpacing.sm)
        }
    }

    private var sheet: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                SheetGrabber()
                    .frame(maxWidth: .infinity)

                JourneySummaryCard(
                    pickup: model.state.draft.pickup,
                    pickupIsChosen: model.state.draft.pickupIsChosen,
                    dropoff: model.state.draft.dropoff,
                    onEdit: onEditRoute
                )

                pickers

                PublicSection(state: model.state, onSelect: { model.selectRoute($0) })

                SectionLabel(key: "booking_private_section")
                privateSection

                // *"Payment chip … private tiers only"*. A bus is not paid for in this app, so the
                // chip is not merely disabled on a public selection — it is absent.
                if !model.state.isPublicSelected {
                    PaymentChipRow(
                        method: model.state.draft.paymentMethod,
                        onChange: { model.setPaymentMethod($0) }
                    )
                }

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                cta
            }
            .padding(MageRideSpacing.md)
        }
        .background(
            MageRideColor.background,
            in: TopRoundedRectangle(radius: MageRideRadius.lg)
        )
    }

    /// The cell's two `seg` controls — *"For Me / Someone"* and *"Person / Package"*, side by side.
    ///
    /// A segmented `Picker` each, which is the cell's own `Δ iOS` clause. Choosing the second option
    /// of either **navigates**: a proxy rider and a parcel each have a screen of their own to fill
    /// in, and the picker is what opens it.
    private var pickers: some View {
        HStack(spacing: MageRideSpacing.xs) {
            Picker("", selection: bookingForBinding) {
                ForEach(BookingFor.allCases, id: \.self) { value in
                    Text(key: value.labelKey).tag(value)
                }
            }
            .pickerStyle(.segmented)
            .labelsHidden()
            .accessibilityLabel(Text(key: "booking_for_label"))

            Picker("", selection: subjectBinding) {
                ForEach(BookingSubject.allCases, id: \.self) { value in
                    Text(key: value.labelKey).tag(value)
                }
            }
            .pickerStyle(.segmented)
            .labelsHidden()
            .accessibilityLabel(Text(key: "booking_subject_label"))
        }
    }

    @ViewBuilder
    private var privateSection: some View {
        if model.state.tiersLoading {
            LoadingRow(messageKey: "booking_estimating")
        } else if model.state.tiers.isEmpty {
            MutedRow(messageKey: "booking_no_tiers")
        } else {
            ForEach(model.state.tiers) { quote in
                TierRow(
                    quote: quote,
                    isSelected: model.state.selection == .privateTier(quote),
                    onSelect: { model.selectTier(quote) }
                )
            }
        }
    }

    /// The one CTA, in whichever of its two forms applies.
    ///
    /// *"Public → Track Route · private → Book Now / Schedule"* is D2' §SCR-PA-009's own sketch.
    /// There is no Schedule under a public route because there is nothing to schedule: a bus runs to
    /// its own timetable and this app does not book a seat on it.
    @ViewBuilder
    private var cta: some View {
        if model.state.isPublicSelected {
            Button(action: onTrackRoute) {
                Text(key: "booking_track_route")
            }
            .buttonStyle(.mageCta)
        } else {
            Button { model.book() } label: {
                Text(key: "booking_book_now")
            }
            .buttonStyle(.mageCta(loading: model.state.booking))
            .disabled(!model.state.canBook)

            OutlinedAction(titleKey: "booking_schedule", action: onSchedule)
        }
    }

    private var bookingForBinding: Binding<BookingFor> {
        Binding(
            get: { model.state.draft.bookingFor },
            set: { value in
                model.setBookingFor(value)
                if value == .someoneElse { onProxyDetails() }
            }
        )
    }

    private var subjectBinding: Binding<BookingSubject> {
        Binding(
            get: { model.state.draft.subject },
            set: { value in
                model.setSubject(value)
                if value == .parcel { onPackageBooking() }
            }
        )
    }
}
